### PR TITLE
fix/enhancement(cli): improve --watch flag

### PR DESCRIPTION
**Pull request by jwx**
The improved --watch flag causes fewer crashes, performs better and is now also available on the build command.

Closes aurelia/cli#293 and aurelia/cli#265.

### CODE REVIEW

File: lib/cli-options.js [hunk 1]
```diff
@@ -5,6 +5,13 @@ exports.CLIOptions = class {
     exports.CLIOptions.instance = this;
   }
 
+  taskName() {
```
Comment (JeroenVinke) [inline lib/cli-options.js]: Would you mind writing some tests for this?

File: lib/resources/tasks/watch.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import transpile from './transpile';
+import processMarkup from './process-markup';
+import processCSS from './process-css';
+import copyFiles from './copy-files';
+
+const debounceWaitTime = 100;
+let isBuilding = false;
+let pendingRefreshPaths = [];
+let watches = {};
+let watchCallback = () => { };
```
Comment (JeroenVinke) [inline lib/resources/tasks/watch.ts]: perhaps `noop` is a better name

File: lib/resources/tasks/run.ts [hunk 3]
```diff
@@ -32,42 +24,31 @@ let serve = gulp.series(
...
 
-    //see if there are static files to be watched
-    if (typeof project.build.copyFiles === 'object') {
-      const files = Object.keys(project.build.copyFiles);
-      gulp.watch(files, refreshCb).on('change', onChangeCb);
-    }
-  };
-};
+function reload() {
+  log("Refreshing browser");
```
Comment (JeroenVinke) [inline lib/resources/tasks/run.ts]: `Refreshing the browser` would be better I think

File: package.json [hunk 4]
```diff
@@ -43,10 +43,12 @@
     "nodemon": "^1.11.0",
     "aurelia-tools": "^0.2.4",
     "babel-eslint": "^7.1.1",
+    "debounce": "^1.0.2",
```
Comment (JeroenVinke) [inline package.json]: do we need these dependencies here or is it sufficient that the CLI adds these deps to the package.json that it creates during `au new`?

File: lib/resources/tasks/build.js [hunk 5]
```diff
@@ -1,26 +1,40 @@
...
-  ),
+    copyFiles),
   writeBundles
 );
 
+let main;
+
+// Only watch if we're the actual task run, not if we're 
+// called from run or other task
+if (CLIOptions.hasFlag("watch") && CLIOptions.taskName() === "build") {
```
Comment (JeroenVinke) [inline lib/resources/tasks/build.js]: if you swap these two expressions then you could remove the comment

File: lib/resources/tasks/build.js [hunk 6]
```diff
@@ -1,26 +1,40 @@
...
+  main = gulp.series(
+    build,
+    (done) => { watch(); done() }
```
Comment (JeroenVinke) [inline lib/resources/tasks/build.js]: nitpicky, but `done()` lacks a semicolon

File: lib/resources/tasks/build.js [hunk 7]
```diff
@@ -1,26 +1,40 @@
...
-import {build} from 'aurelia-cli';
+import watch from './watch';
 import project from '../aurelia.json';
 
-export default gulp.series(
+let build = gulp.series(
   readProjectConfiguration,
-  gulp.parallel(
-    transpile,
+  gulp.parallel(transpile,
```
Comment (JeroenVinke) [inline lib/resources/tasks/build.js]: sorry for being nitpicky, but could you undo this change (and the one at `copyFiles)`) so that the codestyle matches the rest of the gulp tasks?

File: lib/resources/tasks/watch.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+let refresh = debounce(() => {
+  if (isBuilding) {
+    log("Watcher: A build is already in progress, deferring change detection...");
+    return;
+  }
+
+  isBuilding = true;
+
+  let paths = pendingRefreshPaths.splice(0);
+  let refreshTasks = [];
```
Comment (JeroenVinke) [inline lib/resources/tasks/watch.ts]: would `buildTasks` and `pendingBuildTasks` be a better name?

File: lib/resources/tasks/run.js [hunk 9]
```diff
@@ -1,22 +1,10 @@
 import gulp from 'gulp';
 import browserSync from 'browser-sync';
 import historyApiFallback from 'connect-history-api-fallback/lib';
+import { CLIOptions } from 'aurelia-cli';
```
Comment (JeroenVinke) [inline lib/resources/tasks/run.js]: i'm nitpicking again, but could you remove the spaces around `CLIOptions`?

File: lib/cli-options.js (hunk 1 above)
Comment (jwx) [inline lib/cli-options.js]: Sure, I'll see what I can do (that makes sense).

File: lib/resources/tasks/watch.ts (hunk 2 above)
Comment (jwx) [inline lib/resources/tasks/watch.ts]: I don't think so, because it really is the watch callback (it's just initialized to a noop).

File: lib/resources/tasks/watch.ts (hunk 8 above)
Comment (jwx) [inline lib/resources/tasks/watch.ts]: Well, it is the paths (files) pending to be refreshed and it is the tasks that's going to refresh them so I don't really think so. If you feel the names aren't clear enough I could try coming up with some better ones.

File: package.json (hunk 4 above)
Comment (jwx) [inline package.json]: Hm. They might not be needed. I'll check.

File: lib/resources/tasks/build.js (hunk 7 above)
Comment (jwx) [inline lib/resources/tasks/build.js]: Sure. (This is just formatting done by VS Code.)

File: lib/resources/tasks/build.js (hunk 5 above)
Comment (jwx) [inline lib/resources/tasks/build.js]: I added the comment there to emphasize that the build task can be executed by the run task, but I'll happily remove it. For me the order of the conditions doesn't really affect the comment and this order mirrors the one in run.ts, but I'll swap them if you want.

File: lib/resources/tasks/run.js (hunk 9 above)
Comment (jwx) [inline lib/resources/tasks/run.js]: VS Code again. I'm curious, though, why not use spaces within the {} when it's an import statement? They are within {} everywhere else, so is this some convention of the import statement?

Comment (JeroenVinke) [inline lib/resources/tasks/run.js]: We don't use spaces elsewhere in the framework ([example](https://github.com/aurelia/framework/blob/master/src/aurelia.js#L5)) so we should follow that standard

File: lib/resources/tasks/build.js (hunk 5 above)
Comment (JeroenVinke) [inline lib/resources/tasks/build.js]: Yes, please swap them. It's purely for readability, A comment is fine :+1:

File: lib/resources/tasks/watch.ts (hunk 8 above)
Comment (JeroenVinke) [inline lib/resources/tasks/watch.ts]: Right, I see what you mean. When I read "refresh" I thought of refreshing the browser, which happens only once after all changes have been built.

File: lib/resources/tasks/run.js (hunk 9 above)
Comment (jwx) [inline lib/resources/tasks/run.js]: So it's an Aurelia standard? Hm, I wonder if the Aurelia VS Code plugin would be willing to accommodate this so that VS Code's autoformat feature will still work...

Comment (gheoan) [inline lib/resources/tasks/run.js]: An lint rule could be added to https://github.com/aurelia/tools/blob/e457d2d753e0abdc1a8f75d5d9e79ccb464f4f7c/.eslintrc.json in order to standardise the spacing.

**Comment by JeroenVinke** [comment]
@jwx great work! I did encounter one issue when running `au test --watch`:

```
C:\Development\aurelia-cli-watch-flag>au test --watch
C:\Development\aurelia-cli-watch-flag\aurelia_project\tasks\test.js:47
  unit = _gulp2.default.series(_build2.default, _gulp2.default.parallel((0, _run.watch)(_build2.default, onChange), karma));
                                                                                       ^

TypeError: (0 , _run.watch) is not a function
    at Object.<anonymous> (C:/Development/aurelia-cli-watch-flag/aurelia_project/tasks/test.js:29:7)
    at Module._compile (module.js:571:32)
    at loader (C:\Development\cli\node_modules\babel-register\lib\node.js:144:5)
    at Object.require.extensions.(anonymous function) [as .js] (C:\Development\cli\node_modules\babel-register\lib\node.js:154:7)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at process.nextTick (C:\Development\cli\lib\commands\gulp.js:28:43)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickDomainCallback (internal/process/next_tick.js:128:9)
```

Could you look into this? I assume it could be the same logic as `au build --watch`
